packages/marshal/src/marshal.js

```diff
@@ -216,28 +287,84 @@ export const makeMarshal = (
       return val;
     };
 
-    const fullRevive = makeDecodeFromCapData({
+    const decodeRemotableFromSmallcaps = stringEncoding => {
+      assert(stringEncoding.startsWith('$'));
+      // slots: $slotIndex.iface or $slotIndex
+      const i = stringEncoding.indexOf('.');
+      const index = Number(stringEncoding.slice(1, i));
+      // assert(i > 0, X`${rawTree} is not a valid slot encoding`);
+      // i < 0 means there was no iface included
+      const iface = i < 0 ? undefined : stringEncoding.slice(i + 1);
+      return unserializeSlot(index, iface);
+    };
+
+    const decodePromiseFromSmallcaps = stringEncoding => {
+      assert(stringEncoding.startsWith('&'));
+      // slots: $slotIndex.iface or $slotIndex
+      const i = stringEncoding.indexOf('.');
+      const index = Number(stringEncoding.slice(1, i));
```

Comment:
This seems to suffer the same problem. Suggestion:
```js
const makeDecodeSlotFromSmallcaps = prefix => {
  return stringEncoding => {
    assert(stringEncoding.startsWith(prefix));
    // slots: $slotIndex.iface or $slotIndex
    const i = stringEncoding.indexOf('.');
    const slotIndex = Number(stringEncoding.slice(1, i < 0 ? Infinity : i));
    // i < 0 means there was no iface included.
    const iface = i < 0 ? undefined : stringEncoding.slice(i + 1);
    return unserializeSlot(slotIndex, iface);
  };
};
const decodeRemotableFromSmallcaps = makeDecodeSlotFromSmallcaps('$');
const decodePromiseFromSmallcaps = makeDecodeSlotFromSmallcaps('&');
```

Reply:
As above, agreed

Reply:
Done